### PR TITLE
Base Class Library (BCL) module

### PR DESCRIPTION
**Pull request by v-smatveyenka**

### CODE REVIEW

File: 02_C# Fundamentals/SystemWatcherApp/SystemWatcherApp/Program.cs [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+            rules = new List<RuleElement>();
+            startDirectories = new List<string>();
+            watchers = new List<FileSystemWatcher>();
+        }
+
+        static void Main(string[] args)
+        {
+            Console.OutputEncoding = Encoding.UTF8;
+
+            WatcherSectionGroup watcherSectionGroup = ConfigurationManager.OpenExeConfiguration(ConfigurationUserLevel.None).SectionGroups["watcherSection"] as WatcherSectionGroup;
```
Comment (ikolyaskin) [inline 02_C# Fundamentals/SystemWatcherApp/SystemWatcherApp/Program.cs]: I never seen method OpenExeConfiguration, why is this needed? Can't we just do `ConfigurationManager.Section["abc123"]`?

File: 02_C# Fundamentals/SystemWatcherApp/SystemWatcherApp/Program.cs [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+            foreach (ConfigurationSection section in watcherSectionGroup.Sections)
```
Comment (ikolyaskin) [inline 02_C# Fundamentals/SystemWatcherApp/SystemWatcherApp/Program.cs]: Now this is something very weird. Are you cycling through sections here? But don't you already have an object of a class, that has config section mapped onto it? Your `WatcherSectionGroup` even have properties for accessing `RuleSection` and `LocaleSection`, which are never used for some reason.

File: 02_C# Fundamentals/SystemWatcherApp/SystemWatcherApp/Program.cs [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+using SystemWatcherApp.Configs;
+using SystemWatcherApp.Resources;
+
+namespace SystemWatcherApp
+{
+    class Program
+    {
+        readonly static List<string> startDirectories;
+        readonly static List<FileSystemWatcher> watchers;
+        readonly static string defaultDirectory = @"D:\_02_BCL\Default\";
```
Comment (ikolyaskin) [inline 02_C# Fundamentals/SystemWatcherApp/SystemWatcherApp/Program.cs]: Things like this should be in config

File: 02_C# Fundamentals/SystemWatcherApp/SystemWatcherApp/Configs/WatcherSectionGroup.cs [hunk 4]
```diff
@@ -0,0 +1,19 @@
+﻿using System.Configuration;
+
+namespace SystemWatcherApp.Configs
+{
+    public class WatcherSectionGroup : ConfigurationSectionGroup
```
Comment (ikolyaskin) [inline 02_C# Fundamentals/SystemWatcherApp/SystemWatcherApp/Configs/WatcherSectionGroup.cs:5]: Where is `TrackingFoldersSection`?

File: 02_C# Fundamentals/SystemWatcherApp/SystemWatcherApp/Program.cs [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+                {
+                    LocaleSection localeSection = (LocaleSection)section;
+                    switch (localeSection.LocaleCode)
+                    {
+                        case "ru":
+                            currentCulture = new CultureInfo("ru-RU");
+                            break;
+                        default:
+                            currentCulture = new CultureInfo("en-US");
+                            break;
```
Comment (ikolyaskin) [inline 02_C# Fundamentals/SystemWatcherApp/SystemWatcherApp/Program.cs]: For the sake of simplicity you could have the valid culture codes in configuration already. So that instead of this `switch` - `case` you would simply write:
```
currentCulture = new CultureInfo(localeSection.LocaleCode);
```

File: 02_C# Fundamentals/SystemWatcherApp/SystemWatcherApp/Program.cs [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+
+            Thread.CurrentThread.CurrentUICulture = currentCulture;
+
+            Run();
+        }
+
+        public static void Run()
+        {
+            // establish environment
+            {
```
Comment (ikolyaskin) [inline 02_C# Fundamentals/SystemWatcherApp/SystemWatcherApp/Program.cs:67]: Why are these brackets here?

File: 02_C# Fundamentals/SystemWatcherApp/SystemWatcherApp/Program.cs [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+                    }
+                }
+            }
+
+            Thread.CurrentThread.CurrentUICulture = currentCulture;
+
+            Run();
+        }
+
+        public static void Run()
```
Comment (ikolyaskin) [inline 02_C# Fundamentals/SystemWatcherApp/SystemWatcherApp/Program.cs]: We can clearly see different things happening in this method: creating folders, creating file system watcher, running the endless cycle. It is better the keep such things in a separate methods, each of which has it's own specific purpose and is short and easy to understand. You partly did this by separating the method `Run()` from the rest of what is happening in `Main()`, but `Run` is too generic, it doesn't mean anything specific and it still does a number of things not related to each other.

**Review by ikolyaskin — CHANGES_REQUESTED** [review]

File: 02_C# Fundamentals/SystemWatcherApp/SystemWatcherApp/Program.cs (hunk 1 above)
Comment (v-smatveyenka) [reply 02_C# Fundamentals/SystemWatcherApp/SystemWatcherApp/Program.cs]: No, we cannot, because for `ConfigurationSectionGroup` the simpler approach does not work. They suggest using `OpenExeConfiguration` https://docs.microsoft.com/en-us/dotnet/api/system.configuration.configurationsectiongroup?view=netframework-4.8#examples. This is why I have chosen this way

Comment (ikolyaskin) [reply 02_C# Fundamentals/SystemWatcherApp/SystemWatcherApp/Program.cs]: Aaaah you are using SectionGroup... Why? Can we just use ConfigurationSection? Look at how Marina is using it, it looks very simple:
image

Comment (v-smatveyenka) [reply 02_C# Fundamentals/SystemWatcherApp/SystemWatcherApp/Program.cs]: Yes, I can. I have just decided to collect sections in the separate group. Should I give up the `ConfigurationSectionGroup`?

File: 02_C# Fundamentals/SystemWatcherApp/SystemWatcherApp/Program.cs (hunk 2 above)
Comment (v-smatveyenka) [reply 02_C# Fundamentals/SystemWatcherApp/SystemWatcherApp/Program.cs]: I have refactored this

File: 02_C# Fundamentals/SystemWatcherApp/SystemWatcherApp/Program.cs (hunk 3 above)
Comment (v-smatveyenka) [reply 02_C# Fundamentals/SystemWatcherApp/SystemWatcherApp/Program.cs]: Removed it to config

File: 02_C# Fundamentals/SystemWatcherApp/SystemWatcherApp/Configs/WatcherSectionGroup.cs (hunk 4 above)
Comment (v-smatveyenka) [reply 02_C# Fundamentals/SystemWatcherApp/SystemWatcherApp/Configs/WatcherSectionGroup.cs:5]: It has been added

File: 02_C# Fundamentals/SystemWatcherApp/SystemWatcherApp/Program.cs (hunk 5 above)
Comment (v-smatveyenka) [reply 02_C# Fundamentals/SystemWatcherApp/SystemWatcherApp/Program.cs]: I have applied this approach

File: 02_C# Fundamentals/SystemWatcherApp/SystemWatcherApp/Program.cs (hunk 6 above)
Comment (v-smatveyenka) [reply 02_C# Fundamentals/SystemWatcherApp/SystemWatcherApp/Program.cs:67]: removed

File: 02_C# Fundamentals/SystemWatcherApp/SystemWatcherApp/Program.cs (hunk 7 above)
Comment (v-smatveyenka) [reply 02_C# Fundamentals/SystemWatcherApp/SystemWatcherApp/Program.cs]: I have restructured that

File: 02_C# Fundamentals/SystemWatcherApp/SystemWatcherApp/Program.cs (hunk 1 above)
Comment (ikolyaskin) [reply 02_C# Fundamentals/SystemWatcherApp/SystemWatcherApp/Program.cs]: I think we can leave it as is, as you already implemented it, but I think that was an unnecessary complication.

File: 02_C# Fundamentals/SystemWatcherApp/SystemWatcherApp/Program.cs [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+            while (true) ;
+        }
+
+        public static void ApplyConfigurationSettings()
+        {
+            WatcherSectionGroup watcherSectionGroup = ConfigurationManager.OpenExeConfiguration(ConfigurationUserLevel.None).SectionGroups["watcherSection"] as WatcherSectionGroup;
+
+            RuleSection ruleSection = watcherSectionGroup.GeneralSettings;
+            RuleElementCollection ruleElementCollection = ruleSection.Rules;
+            foreach (RuleElement ruleElement in ruleElementCollection)
```
Comment (ikolyaskin) [inline 02_C# Fundamentals/SystemWatcherApp/SystemWatcherApp/Program.cs]: I think that `ruleElementCollection` variable is unnecessary here, as you could simply use `ruleSection.Rules` directly in the `foreach`.

File: 02_C# Fundamentals/SystemWatcherApp/SystemWatcherApp/Program.cs [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+            currentCulture = new CultureInfo(localeSection.LocaleCode);
+
+            TrackingFoldersSection trackingFoldersSection = watcherSectionGroup.TrackingFoldersSettings;
+            TrackingFolderElementCollection trackingFolderElementCollection = trackingFoldersSection.TrackingFolders;
+            foreach (TrackingFolderElement trackingFolderElement in trackingFolderElementCollection)
+            {
+                startDirectories.Add(trackingFolderElement.Address);
+            }
+
+            Thread.CurrentThread.CurrentUICulture = currentCulture;
```
Comment (ikolyaskin) [inline 02_C# Fundamentals/SystemWatcherApp/SystemWatcherApp/Program.cs]: You use the variable `currentCulture` in the line 58, then you do something else not related to the culture, and then you again do something with this variable on line 67. It is better to keep things related to one subject in one place. I don't see any reason why you wouldn't put this line right after putting the value in `currentCulture`.

Also, I'm not sure I understand, why do we even need this static field in Program for current culture? If you are setting it to the current thread, maybe it would be better to use it from there, rather than having it as static thing in Program?

**Review by ikolyaskin — CHANGES_REQUESTED** [review]

File: 02_C# Fundamentals/SystemWatcherApp/SystemWatcherApp/Program.cs (hunk 8 above)
Comment (v-smatveyenka) [reply 02_C# Fundamentals/SystemWatcherApp/SystemWatcherApp/Program.cs]: fixed

File: 02_C# Fundamentals/SystemWatcherApp/SystemWatcherApp/Program.cs (hunk 9 above)
Comment (v-smatveyenka) [reply 02_C# Fundamentals/SystemWatcherApp/SystemWatcherApp/Program.cs]: the variable `currentCulture` has been deleted

**Review by ikolyaskin — APPROVED** [review]
